File: plugins/cbre-il/skills/cbre-property-longlist/evals/f32_rent_basis_test.py
#!/usr/bin/env python3
"""f32_rent_basis_test.py - a Total rent figure always says what area it was computed on. (SEAM-16)

WHAT WAS WRONG. v41 (F26) made every "Total GLA" label read the source's own stated total when
one was surfaced, which was right: a deck's printed total routinely includes a gatehouse, a plant
room or a hub office that the two summed fields do not hold. But totalAnnualRent() still
multiplies the rate by the SUMMED warehouse + office, so the modal could print a Total GLA of
45,649 a few rows above a Total annual rent computed on 42,000, and a reader dividing one by the
other got a rate no source quoted, with nothing on the page to explain the gap. Every rent in the
run that exposed it was unknown, so nothing was visibly wrong; this is contract-driven.

THE RESOLUTION, and what it is not. The rent STAYS on the summed lettable area (pricing a stated
total that includes space nobody pays warehouse rent on would overstate, and this is money). What
changed is that the figure now carries its BASIS beside it: an area x rate formula (split into two
terms when a separate office rate applies) as a .rent-basis sub-line under the annual figure in
the modal and in the compare matrix. It is a formula of numbers and units rather than a sentence,
so no i18n key was added. It prints the COMPONENTS, "(40,000 + 2,000) sq ft x rate", not their
sum: the reader can match each term to the Warehouse area and Office area rows directly above and
see which fields were priced, and the summed figure itself still appears nowhere on the page,
which is what f26_gla_single_derivation_test pins for the GLA label. An unknown office area is
simply absent from the formula, which IS the disclosure; an unknown rent still omits the row,
exactly as before. No money figure changed.

THE INVARIANT THIS PINS (the shape A1 used for the GLA twin): rentBasis() is the ONE place the
inputs are qualified, and it has exactly TWO consumers, the multiplication (totalAnnualRent) and
the printed basis (rentBasisStr), so the two cannot drift apart; and every Total rent render site
reaches the page through totalRentHTML(), the only caller of totalRentStr(), so a figure cannot be
printed without its basis by a later edit that calls the string helper directly.

Offline. Needs node to execute the chrome (as f26_gla_single_derivation_test does).
"""
from __future__ import annotations

import hashlib
import json
import re
import shutil
import subprocess
import sys
import tempfile
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
HELPERS = ROOT / "helpers"
sys.path.insert(0, str(HELPERS))
import _common as C  # noqa: E402
import build_dashboard as BD  # noqa: E402

PX = ("data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0"
      "lEQVR42mNkYPhfDwAChwGA60e6kgAAAABJRU5ErkJggg==")
FAILS: list[str] = []


def ck(ok, msg):
    print(("  [PASS] " if ok else "  [FAIL] ") + msg)
    if not ok:
        FAILS.append(msg)


def _canon():
    base = {"country": "GB", "developer": "Dev", "city": "Sometown", "status": "Available",
            "photo": PX, "gallery": [PX], "lat": 52.49, "lng": -0.69, "areaUnit": "sq ft",
            "rentUnit": "\u00a3/sq ft/yr", "warehouseRent": "\u00a39.50 / sq ft / year",
            "warehouseRentVal": 9.5}
    props = [
        # 40,000 + 2,000 = 42,000 summed vs 45,649 printed -> the GLA row reads the stated total,
        # the rent is computed on 42,000, and the basis line must SAY 42,000
        dict(id=1, park="Disagrees", warehouseArea=40000, officeArea="2000", officeAreaVal=2000),
        # a separate office rate -> two terms, each area at its own rate
        dict(id=2, park="Split", warehouseArea=30000, officeArea="1000", officeAreaVal=1000,
             officeRent="\u00a312.00 / sq ft / year", officeRentVal=12.0),
        # office area unknown -> the warehouse area ALONE is the basis, and the formula shows it
        dict(id=3, park="OfficeUnknown", warehouseArea=20000, officeArea="tbd"),
        # rent unknown -> no total, no basis, no row (tbd is first-class, never smoothed over)
        dict(id=4, park="RentUnknown", warehouseArea=20000, warehouseRent="tbd", warehouseRentVal=None),
    ]
    stated = {"1": {"value": 45649, "unit": "sq ft", "source_file": "t.xlsx", "locator": "S!r2"}}
    out = []
    for p in props:
        d = dict(base, **p)
        if d.get("warehouseRentVal") is None:
            d.pop("warehouseRentVal", None)
        out.append(d)
    return {"meta": {"client": "RentBasis",
                     "units": {"area": "sq ft", "rent": "\u00a3/sq ft/yr"},
                     "hero": {"topbar_meta": "", "eyebrow": "", "title_html": "",
                              "lede": "", "footer_copyright": ""},
                     "statedTotals": stated},
            "pois": [], "regions": {},
            "properties": out}


def main() -> int:
    tpl = C.load_template()

    print("== structure: ONE qualification of the inputs, TWO consumers (the money and its printed basis) ==")
    ck("function rentBasis(p){" in tpl, "rentBasis() exists")
    n_rb = tpl.count("rentBasis(p)")
    ck(n_rb == 3,
       f"rentBasis(p) occurs exactly 3 times: its definition, totalAnnualRent() and rentBasisStr() ({n_rb})")
    tar = tpl.split("function totalAnnualRent(p){", 1)[-1].split("\n}", 1)[0]
    ck("const b = rentBasis(p);" in tar, "totalAnnualRent() takes its inputs FROM rentBasis()...")
    ck("p.warehouseRentVal" not in tar and "p.officeAreaVal" not in tar and "p.officeRentVal" not in tar,
       "...and qualifies none of them itself (no second copy of the guards)")
    ck("(wa + oa) * wr" in tar and "wa * wr + oa * or" in tar,
       "the multiplication is still summed warehouse + office, split per rate when an office rate exists")
    ck("glaVal(" not in tar and "statedTotal(" not in tar,
       "money is never re-derived from the stated total (v41's rule, kept)")
    rbs = tpl.split("function rentBasisStr(p){", 1)[-1].split("\n}", 1)[0]
    ck("const b = rentBasis(p);" in rbs and "if(b.split)" in rbs
       and "term(fmt(b.wa), b.wr)" in rbs and "term(fmt(b.oa), b.or)" in rbs
       and '"(" + fmt(b.wa) + " + " + fmt(b.oa) + ")"' in rbs,
       "rentBasisStr() prints exactly the components totalAnnualRent() multiplies, on the same split test")
    ck("b.wa + b.oa" not in rbs, "the basis prints the components, never their sum (the GLA label owns no summed figure)")
    after = tpl.split("function totalAnnualRent(", 1)[-1][:600]
    ck("function rentBasis(p){" in after and "p.officeAreaVal" in after,
       "rentBasis() is declared directly after totalAnnualRent() (hoisted; two older evals read the guards there)")

    print()
    print("== structure: every Total rent render site reaches the page through totalRentHTML() ==")
    ck("function totalRentHTML(p, monthly){" in tpl, "totalRentHTML() exists")
    n_str = tpl.count("totalRentStr(p")
    ck(n_str == 2,
       f"totalRentStr(p...) occurs exactly twice: its definition and totalRentHTML()'s call ({n_str})")
    ck(re.search(r"totalRentStr\(p,\s*(false|true)\)", tpl) is None,
       "no render site calls totalRentStr() directly (a figure cannot ship without its basis)")
    annual_lines = [ln for ln in tpl.splitlines() if "row_total_annual_rent" in ln and "T(" in ln]
    ck(len(annual_lines) == 2 and all("totalRentHTML(p,false)" in ln for ln in annual_lines),
       f"the Total annual rent label is rendered on exactly two surfaces (modal, compare), both via totalRentHTML(p,false) ({len(annual_lines)})")
    monthly_lines = [ln for ln in tpl.splitlines() if "row_total_monthly_rent" in ln and "T(" in ln]
    ck(len(monthly_lines) == 2 and all("totalRentHTML(p,true)" in ln for ln in monthly_lines),
       f"the Total monthly rent label likewise, via totalRentHTML(p,true) ({len(monthly_lines)})")
    ck(tpl.count('<small class="rent-basis">') == 1 and ".rent-basis{" in tpl,
       "the .rent-basis sub-line is emitted from ONE place and has a CSS rule")
    trh = tpl.split("function totalRentHTML(p, monthly){", 1)[-1].split("\n}", 1)[0]
    ck("monthly ? null : rentBasisStr(p)" in trh,
       "the basis rides on the ANNUAL figure; the monthly figure (annual / 12 by its label) carries none")

    print()
    print("== no new i18n key: the basis is numbers and units, not authored prose ==")
    ck(re.search(r"""T\(["']rent_basis["']\)""", tpl) is None and "rent_basis" not in json.dumps(
        __import__("i18n").EN), "no rent_basis i18n key exists or is referenced")

    print()
    print("== the version bump and the hash chain the template contract requires ==")
    label = (ROOT / "assets" / "VERSION").read_text(encoding="utf-8").splitlines()[0].strip()
    ck(int(re.sub(r"\D", "", label) or 0) >= 42, f"assets/VERSION is at/after v42 ({label})")
    expected = hashlib.sha256(tpl.encode("utf-8")).hexdigest()
    recorded = C.load_version().get("chrome_sha256", "")
    ck(recorded == expected,
       f"VERSION's chrome_sha256 is the live template's text hash (recorded {recorded[:12]}, computed {expected[:12]})")

    print()
    print("== a real build, executed ==")
    with tempfile.TemporaryDirectory() as td:
        d = Path(td)
        cp, hp = d / "c.json", d / "b.html"
        cp.write_text(json.dumps(_canon()), encoding="utf-8")
        BD.build(cp, hp)
        built = hp.read_text(encoding="utf-8")
        again, _ = BD.render(C.load_canonical(cp))
        ck(again == built, "re-rendering the same canonical is byte-identical (validate-html's floor)")
        node = shutil.which("node") or r"C:\Users\TBaaij\nodejs\node.exe"
        mjs = Path(__file__).with_suffix(".mjs")
        if not Path(node).exists() and not shutil.which("node"):
            ck(False, "node is required to execute the chrome (install node or add it to PATH)")
        else:
            r = subprocess.run([node, str(mjs), str(hp)], capture_output=True, text=True,
                               encoding="utf-8", errors="replace")
            print((r.stdout or "").rstrip())
            if r.returncode != 0:
                print((r.stderr or "").rstrip())
            ck(r.returncode == 0, "the EXECUTED basis assertions pass (see above)")

    print()
    if FAILS:
        print(f"F32 RENT BASIS TEST: FAIL ({len(FAILS)})")
        return 1
    print("F32 RENT BASIS TEST: PASS")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: plugins/cbre-il/skills/cbre-property-longlist/evals/f32_rent_basis_test.mjs
// evals/f32_rent_basis_test.mjs - execute the REAL chrome from a BUILT dashboard and prove that
// every Total annual rent figure on the page carries the area x rate it was computed on, that
// the printed basis multiplies back to the printed figure, and that no money figure changed.
// (SEAM-16)  Offline; no npm deps.  Usage: node f32_rent_basis_test.mjs <built_html_path>
import fs from 'node:fs';
import vm from 'node:vm';

const htmlPath = process.argv[2];
if (!htmlPath) { console.error('usage: node f32_rent_basis_test.mjs <built_html>'); process.exit(2); }
const html = fs.readFileSync(htmlPath, 'utf8');

const scripts = [...html.matchAll(/<script(?![^>]*\bsrc=)[^>]*>([\s\S]*?)<\/script>/gi)].map(m => m[1]);
const NAMES = ['PROPS', 'BLANK', 'glaVal', 'glaStr', 'rentBasis', 'rentBasisStr', 'totalAnnualRent', 'totalRentStr',
               'totalRentHTML', 'detailHTML', 'compareHTML', 'AREA_UNIT', 'T'];
const code = scripts.join('\n;\n') + '\n;\n' +
  NAMES.map(n => `__capture__(${JSON.stringify(n)}, typeof ${n} !== "undefined" ? ${n} : undefined);`).join('\n') + '\n';

// See modal_render_test.mjs for why this sandbox is shaped the way it is.
const sink = new Proxy(function () {}, {
  get: (_t, p) => (p === Symbol.toPrimitive || p === 'toString' || p === 'valueOf') ? () => '' : sink,
  apply: () => sink, construct: () => sink, has: () => true,
});
const target = { console };
for (const name of Object.getOwnPropertyNames(globalThis)) {
  if (!(name in target)) { try { target[name] = globalThis[name]; } catch { /* ignore */ } }
}
target.globalThis = target;
target.__capture__ = (name, val) => { target[name] = val; };
const ctx = vm.createContext(new Proxy(target, {
  get: (t, p) => (p in t ? t[p] : sink),
  has: () => true,
}));
try { vm.runInContext(code, ctx, { filename: 'built.inline.js' }); }
catch (e) { console.error('FAIL: template script threw during eval:', e && e.message); process.exit(1); }

const { PROPS: props, BLANK, glaVal, glaStr, rentBasis, rentBasisStr, totalAnnualRent, totalRentStr, totalRentHTML,
        detailHTML, compareHTML, AREA_UNIT, T } = target;
for (const [n, f] of Object.entries({ glaVal, glaStr, rentBasis, rentBasisStr, totalAnnualRent, totalRentStr,
                                      totalRentHTML, detailHTML, T })) {
  if (typeof f !== 'function') { console.error(`FAIL: could not capture ${n}() from the built chrome`); process.exit(1); }
}
if (!Array.isArray(props)) { console.error('FAIL: PROPS not captured'); process.exit(1); }

const fails = [];
const ck = (ok, label) => { console.log((ok ? '  ok   ' : '  FAIL ') + label); if (!ok) fails.push(label); };
const esc = s => s.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
const ANNUAL = T('row_total_annual_rent'), MONTHLY = T('row_total_monthly_rent'), GLA = T('row_total_gla');
// the modal row carrying a label -> its full spec-v innerHTML (null when the row is omitted)
const rowHtml = (h, label) => {
  const m = h.match(new RegExp(`<div class="spec-k">${esc(label)}</div><div class="spec-v">([\\s\\S]*?)</div></div>`));
  return m ? m[1] : null;
};
const num = s => Number(String(s).replace(/[^0-9.]/g, ''));
// parse every "(A [+ B]) u × C r / u" term back out of a basis line and re-multiply
const TERM_RX = new RegExp(`\\(?([\\d,.]+)(?: \\+ ([\\d,.]+))?\\)? ${esc(AREA_UNIT)} × \\S+ ([\\d,.]+) \\/ [^+]+?(?= \\+ |$)`, 'g');
const basisTotal = basis => {
  const terms = [...String(basis).matchAll(TERM_RX)];
  if (!terms.length) return NaN;
  return terms.reduce((acc, m) => acc + (num(m[1]) + (m[2] ? num(m[2]) : 0)) * num(m[3]), 0);
};

const byPark = n => props.find(p => p.park === n);
const disagrees = byPark('Disagrees'), split = byPark('Split'), ounk = byPark('OfficeUnknown'), runk = byPark('RentUnknown');
if (!disagrees || !split || !ounk || !runk) { console.error('FAIL: fixture properties missing'); process.exit(1); }

console.log('-- the figure and its basis, on a property whose GLA row reads a LARGER stated total --');
const mD = detailHTML(disagrees);
const annualD = rowHtml(mD, ANNUAL);
ck(rowHtml(mD, GLA) === '45,649 sq ft', `the GLA row prints the stated total (${JSON.stringify(rowHtml(mD, GLA))})`);
ck(annualD !== null && /^£ 399,000 \/ yr<small class="rent-basis">/.test(annualD),
   `the Total annual rent row prints the figure then a .rent-basis sub-line: ${JSON.stringify(annualD)}`);
const basisD = (annualD && annualD.match(/<small class="rent-basis">([^<]*)<\/small>/) || [])[1] || null;
ck(basisD === `(40,000 + 2,000) ${AREA_UNIT} × £ 9.50 / sq ft`,
   `the basis names the two lettable COMPONENTS at the quoted rate (${JSON.stringify(basisD)})`);
ck(basisD !== null && !/45,649/.test(basisD) && totalAnnualRent(disagrees) === 42000 * 9.5,
   'the rent is NOT computed on the stated total, and the basis does not claim it was');
ck(!/42,000/.test(mD), 'the summed figure itself appears nowhere in the modal (f26\'s pin for the GLA label holds)');
ck(Math.abs(basisTotal(basisD) - totalAnnualRent(disagrees)) < 0.5,
   `the printed basis multiplies back to the printed figure (${basisTotal(basisD)})`);
ck(annualD === totalRentHTML(disagrees, false), 'the modal row IS totalRentHTML(): no second composition');
ck(rowHtml(mD, MONTHLY) === '£ 33,250 / mo',
   `the monthly figure is the annual / 12 with no basis of its own (${JSON.stringify(rowHtml(mD, MONTHLY))})`);
ck(totalRentStr(disagrees, false) === '£ 399,000 / yr' && totalRentStr(disagrees, true) === '£ 33,250 / mo',
   'the money strings are byte-identical to v41: no money figure changed');

console.log('\n-- a separate office rate: two terms, each area at its own rate --');
const mS = detailHTML(split);
const basisS = (rowHtml(mS, ANNUAL) || '').match(/<small class="rent-basis">([^<]*)<\/small>/);
ck(basisS && basisS[1] === `30,000 ${AREA_UNIT} × £ 9.50 / sq ft + 1,000 ${AREA_UNIT} × £ 12.00 / sq ft`,
   `split basis: ${JSON.stringify(basisS && basisS[1])}`);
ck(totalAnnualRent(split) === 30000 * 9.5 + 1000 * 12 && Math.abs(basisTotal(basisS ? basisS[1] : '') - totalAnnualRent(split)) < 0.5,
   `split total ${totalAnnualRent(split)} equals the re-multiplied basis`);
ck(rentBasis(split).split === true && rentBasis(disagrees).split === false,
   'rentBasis().split decides BOTH the multiplication and the printed shape');

console.log('\n-- an unknown office area is absent from the formula: that absence is the disclosure --');
const basisO = (rowHtml(detailHTML(ounk), ANNUAL) || '').match(/<small class="rent-basis">([^<]*)<\/small>/);
ck(basisO && basisO[1] === `20,000 ${AREA_UNIT} × £ 9.50 / sq ft`,
   `warehouse area alone, no guessed office (${JSON.stringify(basisO && basisO[1])})`);
ck(totalAnnualRent(ounk) === 20000 * 9.5, 'and the money matches it');

console.log('\n-- an unknown rent stays unknown: no total, no basis, no figure --');
const mR = detailHTML(runk);
// v45: the commercial block always prints, so the two rows are PRESENT and carry the blank
// token. What this eval has always guarded is that no FIGURE and no BASIS is invented for a
// property with no stated rate, and that is asserted below on the helpers themselves.
ck(rowHtml(mR, ANNUAL) === BLANK && rowHtml(mR, MONTHLY) === BLANK,
   'both Total rent rows print the blank token, not a figure');
ck(!/rent-basis/.test(mR), 'no basis sub-line anywhere in that modal');
ck(totalAnnualRent(runk) === null && rentBasis(runk) === null && rentBasisStr(runk) === null && totalRentHTML(runk, false) === null,
   'every helper returns null rather than a figure');

console.log('\n-- the helpers refuse what totalAnnualRent() always refused --');
ck(rentBasis({ warehouseRentVal: 'tbd', warehouseArea: 1000 }) === null
   && rentBasis({ warehouseRentVal: 5, warehouseArea: 0 }) === null
   && rentBasis({ warehouseRentVal: -1, warehouseArea: 1000 }) === null,
   'sentinel, zero and negative inputs -> null');
const b = rentBasis({ warehouseRentVal: 5, warehouseArea: 1000, officeAreaVal: 100, officeRentVal: -2 });
ck(b && b.oa === 100 && b.or === null && b.split === false,
   'a non-positive office rate is dropped and the office area rides on the warehouse rate');

if (typeof compareHTML === 'function') {
  let cmp = null;
  try { cmp = compareHTML([disagrees, split]); } catch (e) { cmp = null; }
  if (typeof cmp === 'string' && cmp.length) {
    ck(/<small class="rent-basis">\(40,000 \+ 2,000\) /.test(cmp) && /<small class="rent-basis">30,000 /.test(cmp) && !/42,000/.test(cmp),
       'the COMPARE matrix prints each property\'s basis under its annual figure');
  } else {
    console.log('  skip compareHTML needs a DOM here; its row is pinned structurally by the Python driver');
  }
}

console.log('STATUS:', fails.length ? 'BLOCKED' : 'ALL-PASS');
process.exit(fails.length ? 1 : 0);
